WCSimFLOWER.cpp

```diff
@@ -0,0 +1,441 @@
+{
+  if(!name.compare("HyperK") || !name.compare("HyperK_40perCent"))
+    return kHyperK40;
+  else if(!name.compare("SuperK"))
+    return kSuperK;
+  else if (!name.compare("HyperK_20perCent"))
+    return kHyperK20;
+  cerr << "FromString() Unknown detector name " << name << endl;
+  exit(-1);
+}
+
+float WCSimFLOWER::GetEnergy(std::vector<float> times, std::vector<int> tubeIds, float * vertex)
+{
+  assert(times.size() == tubeIds.size());
+  fNDigits = times.size();
+  //resize the vector (shouldn't reduce capacity, but increases capacity if required)
+  fDistance           .resize(fNDigits);
+  fTimesCorrected      .resize(fNDigits);
+  fTimesCorrectedSorted.resize(fNDigits);
+  fTimes              .resize(fNDigits);
+  fTubeIds            .resize(fNDigits);
+
+  fTimes   = times;
+  fTubeIds = tubeIds;
+  for(int i = 0; i < 3; i++)
+    fVertex[i]  = vertex[i];
+
+  CorrectHitTimes();
+  FindMaxTimeInterval();
+  GetNEff();
+  CorrectEnergy();
+
+  return fERec;
+}
+
+void WCSimFLOWER::CorrectHitTimes()
+{
+  // Correct all hit times for time-of-flight between vertex and PMT
+  float x, y, z;
+  WCSimRootPMT pmt;
+  for (unsigned int i = 0; i < fNDigits; i++) {
+    pmt = fGeom->GetPMT(fTubeIds[i] - 1);
+    x = pmt.GetPosition(0);
+    y = pmt.GetPosition(1);
+    z = pmt.GetPosition(2);
+    fDistance[i] = sqrt(pow((x - fVertex[0]), 2) +
+			pow((y - fVertex[1]), 2) +
+			pow((z - fVertex[2]), 2));
+    fTimesCorrected[i] = fTimes[i] - (fDistance[i] / fLightGroupSpeed);
+  }
+
+  // Sort a copy of tCorrected, so tCorrected stays in the same order as other vectors
+  fTimesCorrectedSorted = fTimesCorrected;
+  std::sort(fTimesCorrectedSorted.begin(), fTimesCorrectedSorted.end());
+}
+
+void WCSimFLOWER::FindMaxTimeInterval()
+{
+  // Find the x ns interval with the highest number of hits, and its start time
+  fNMaxShort = 0;
+  fNMaxLong  = 0;
+  int nshort, nlong;
+  float time;
+  for (unsigned int i = 0; i < fNDigits; i++) {
+    // count hits within 50 and 100 ns after each entry
+    nshort = 0;
+    nlong  = 0;
+    time   = fTimesCorrectedSorted[i];
+    for (unsigned int j = i; j < fNDigits; j++) {
+      if (fTimesCorrectedSorted[j] < time + fLongDuration) {
+	nlong++;
+	if (fTimesCorrectedSorted[j] < time + fShortDuration)
+	  nshort++;
+      }
+      else
+	break;
+    }//j
+
+    if (nshort > fNMaxShort) {
+      fNMaxShort = nshort;
+      fNMaxLong  = nlong;
+      fStartTime = time;
+    }
+  }//i
+
+  unsigned int j = 0;
+  fDistanceShort.resize(fNMaxShort);
+  fTubeIdsShort .resize(fNMaxShort);
+  for (unsigned int i = 0; i < fNDigits; i++) {
+    if(fTimesCorrected[i] >= fStartTime && fTimesCorrected[i] < (fStartTime + fShortDuration)) {
+      fDistanceShort[j] = fDistance[i];
+      fTubeIdsShort [j] = fTubeIds [i];
+      if(fTubeIdsShort[j] <= 0 || fTubeIdsShort[j] > fNPMTs)
+	cerr << "WTF?! TubeId short has picked up an invalid ID" << fTubeIdsShort[j] << endl;
```

Comment:
What’s going on here? First of all, do tube IDs actually go from 1 to n instead of 0 to n-1? And second, how would such an invalid ID appear? (At first glance, I’d assume this would only happen if we use the wrong detector configuration here in FLOWER (e.g. use `HyperK_20perCent` to reconstruct a root file simulated with the `HyperK_40perCent` configuration). If that’s the case, the error message could be more helpful.)

Reply:
Yes they do. It's weird and confusing I know...
I've made the error message better